File: src/LongestPalindromeSubstring.java
//Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.
//
//        Example:
//
//        Input: "babad"
//
//        Output: "bab"
//
//        Note: "aba" is also a valid answer.
//
//
//        Example:
//
//        Input: "cbbd"
//
//        Output: "bb"

public class LongestPalindromeSubstring {

    // Solution 1. Using dynamic programming. Store the palindromes of size 1 -> s.length().
    // Create a 2-d array, if palindromes[i][j] = true, the substring(i, j) is a palindrome.
    // That way, when we are checking the substring of size k (index from ith -> jth) to be a palindrome or not,
    // we just need to check two things: 1) is charAt(i) == charAt(j)? 2) Is palindromes[i+1][j-1] = true?
    // If both of them are true, we know that the checking substring is a palindrome. These checking take O(1) time

    // ============ Complexity ==============
    // The whole algorithm takes O(n^2) time for checking palindrome of length from 1 -> s.length()
    // and checking the all the string of that length
    //
    // The algorithm has a space complexity of O(n^2)

    public static String longestPalindrome(String s) {
        boolean[][] palindromes = new boolean[s.length()][s.length()];

        int startingIndexPalin = 0, maxLengthPalin = 1;

        for (int i = 0; i < s.length(); i++) {
            palindromes[i][i] = true;
        }

        for (int i = 0; i < s.length() - 1; i++) {
            if (s.charAt(i) == s.charAt(i+1)) {
                palindromes[i][i+1] = true;
                startingIndexPalin = i;
                maxLengthPalin = 2;
            }
        }

        for (int l = 3; l <= s.length(); l++) {
            for (int begin = 0; begin <= s.length() - l; begin++) {
                int end = begin + l - 1;
                if (s.charAt(begin) == s.charAt(end) && palindromes[begin+1][end-1] == true) {
                    palindromes[begin][end] = true;
                    startingIndexPalin = begin;
                    maxLengthPalin = l;
                }
            }
        }

        return s.substring(startingIndexPalin, startingIndexPalin+maxLengthPalin);
    }

    public static void main(String[] args) {
        System.out.println("Input = babad");
        System.out.println("Output = " + LongestPalindromeSubstring.longestPalindrome("babad"));
        System.out.println("Input = cbbd");
        System.out.println("Output = " + LongestPalindromeSubstring.longestPalindrome("cbbd"));
    }
}

File: python/longest_palindrome_substring.py
'''
Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.

Example 1:

Input: "babad"
Output: "bab"
Note: "aba" is also a valid answer.
Example 2:

Input: "cbbd"
Output: "bb"
'''

class Solution:

    def longestPalindrome(self, s: str) -> str:
        """[summary]
        
        Parameters
        ----------
        s : str
            [description]
        
        Returns
        -------
        str
            [description]
        """
        if s is None or len(s) < 2:
            return s
        result = s[:1]
        for i in range(len(s)):
            l1 = self.extendTheCenter(s, i, i)
            l2 = self.extendTheCenter(s, i, i+1)
            l = l1 if len(l1) > len(l2) else l2
            if len(l) > len(result):
                result = l
        return result

    def extendTheCenter(self, s: str, left: int, right: int) -> int:
        """Get longest length of palindrome expanding from current left and right indices
        
        Parameters
        ----------
        s : str
            [description]
        left : int
            [description]
        right : int
            [description]
        
        Returns
        -------
        int
            length of longest palindrome from left and right
        """
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        return s[left+1:right]

    def longestPalindrome2(self, s: str) -> str:
        """Dynamic programming implementation
        
        Parameters
        ----------
        s : str
            [description]
        
        Returns
        -------
        str
            [description]
        """
        length = len(s)

        # dp[i][j] = True if s[i] -> s[j] is a palindrome otherwise False
        dp = [[False for _ in range(length)] for _ in range(length)]
        dp[length-1][length-1] = True
        for cell in range(length - 1):
            dp[cell][cell] = True
            dp[cell][cell+1] = s[cell] == s[cell+1]

        longestLength = 1

        startIndex = endIndex = 0

        for i in range(length):
            for j in range(i+1, length):
                if s[i] == s[j] and dp[i+1][j-1] is True:
                    dp[i][j] = True
                    newPalindromeLength = j - i + 1
                    if longestLength < newPalindromeLength:
                        startIndex, endIndex = i, j
                        longestLength = newPalindromeLength
            
        return s[startIndex:endIndex+1]